src/vs/platform/actions/common/offlineMode.ts

```diff
@@ -0,0 +1,159 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import { localize } from 'vs/nls';
+import { Action } from 'vs/base/common/actions';
+import { TPromise } from 'vs/base/common/winjs.base';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { INotificationService, Severity } from 'vs/platform/notification/common/notification';
+import { MenuRegistry, MenuId } from 'vs/platform/actions/common/actions';
+import { ContextKeyExpr } from 'vs/platform/contextkey/common/contextkey';
+import { IStorageService, StorageScope } from 'vs/platform/storage/common/storage';
+import { CommandsRegistry } from 'vs/platform/commands/common/commands';
+import { IInstantiationService } from 'vs/platform/instantiation/common/instantiation';
+
+export const offlineModeSetting = 'workbench.enableOfflineMode';
+export const unSupportedInOfflineModeMsg = localize('offlineModeUnsupportedFeature', "This feature is not supported in offline mode");
+
+export class EnableOfflineMode extends Action {
+	static readonly ID = 'workbench.action.enableOfflineMode';
+	static LABEL = localize('enableOfflineMode', 'Enable Offline Mode');
+
+	private readonly disclaimerStorageKey = 'workbench.offlineMode.disclaimer.dontShowAgain';
+
+	constructor(
+		id: string = EnableOfflineMode.ID,
+		label: string = EnableOfflineMode.LABEL,
+		@IConfigurationService private configurationService: IConfigurationService,
+		@IStorageService private storageService: IStorageService,
+		@INotificationService private notificationService: INotificationService
+	) {
+		super(id, label);
+		this.updateEnabled();
+		this.configurationService.onDidChangeConfiguration(e => {
+			if (e.affectsConfiguration(offlineModeSetting)) {
+				this.updateEnabled();
+			}
+		});
+	}
+
+	private updateEnabled() {
+		this.enabled = this.configurationService.getValue(offlineModeSetting) !== true;
+	}
+
+	run(): TPromise<any> {
+		if (this.storageService.getBoolean(this.disclaimerStorageKey, StorageScope.GLOBAL, false) === false) {
+			this.notificationService.prompt(Severity.Info, localize('offlineModeDisclaimer', "VS Code cannot stop extensions from making network requests in offline mode. If extensions make such requests, please log an issue against them."), [
```

Comment:
Is there a way for extensions to check if the user in offline mode?

Reply:
@mjbvz Yes, they can look at the value for the `workbench.enableOfflineMode` setting. I havent yet gone through built-in extensions to ensure they are able to comply with the offline mode. That will be another PR.

Reply:
Ok, so that will added to `vscode.d.ts`?

Reply:
Read the setting value directly